Port from ORT's `Result` to Kotlin's `Result`

sschuberth opened the pull request:
Please have a look at the individual commit messages for the details.

mnonnenmacher commented on scanner/src/funTest/kotlin/storages/AbstractStorageFunTest.kt:
```diff
@@ -145,12 +147,18 @@ abstract class AbstractStorageFunTest(vararg listeners: TestListener) : WordSpec
-                readResult.shouldBeSuccess()
-                readResult.result should beEmpty()
+                readResult.shouldBeSuccess {
+                    it.shouldNotBeNull {
```
Why is this sometimes required and sometimes not?

sschuberth replied:
It depends on whether a member of `it` or only `it` itself is accessed. I've fixed this in Kotest, see https://github.com/kotest/kotest/pull/2762.

mnonnenmacher commented on scanner/src/main/kotlin/PathScanner.kt:
```diff
@@ -303,19 +299,16 @@ abstract class PathScanner(
-                filteredResult.copy(summary = summary)
-            }
-        }
+        return storageResult.map { filteredResult }.recover {
```
Could this use `getOrElse { ... }` instead of `recover { ... }.getOrThrow()`?

mnonnenmacher commented on scanner/src/main/kotlin/ScanResultsStorage.kt:
```diff
@@ -372,20 +375,14 @@ abstract class ScanResultsStorage : PackageBasedScanStorage {
+        map { scanResults ->
+            scanResults.filter { it.provenance == nestedProvenance.root }
+                .map { it.toNestedProvenanceScanResult(nestedProvenance) }
+        }.getOrThrow()
```
Is the encapsulated exception always a `ScanStorageException`?

sschuberth replied:
I believe so. In my porting, I payed attention to never throw anything else than `ScanStorageException`.

mnonnenmacher commented on scanner/src/main/kotlin/ScanResultsStorage.kt:
```diff
@@ -372,20 +375,14 @@ abstract class ScanResultsStorage : PackageBasedScanStorage {
-            val result = add(pkg.id, scanResult)
-
-            if (result is Failure) throw ScanStorageException(result.error)
+            add(pkg.id, scanResult).getOrThrow()
```
Is the encapsulated exception always a `ScanStorageException`?

sschuberth replied:
I believe so. In my porting, I payed attention to never throw anything else than `ScanStorageException`.

MarcelBochtler commented on scanner/src/main/kotlin/PathScanner.kt:
```diff
@@ -303,19 +299,16 @@ abstract class PathScanner(
+            val issues = scanSummary.issues + issue
+            val summary = scanSummary.copy(issues = issues)
+            filteredResult.copy(summary = summary)
+        }.getOrThrow()
```
Isn't `getOrThrow()` behaving differently from the existing code?
Wouldn't `getOrElse()` be the correct equivalent instead of `recover` and `getOrThrow()`?

sschuberth replied:
You're right, `getOrElse()` clearly fits better here. Will change.

mnonnenmacher commented on utils/test/src/main/kotlin/Extensions.kt:
```diff
@@ -20,21 +20,13 @@
 
 package org.ossreviewtoolkit.utils.test
 
-import io.kotest.assertions.withClue
```
Commit message typo: "introd**u**ced"

sschuberth commented:
@mnonnenmacher I've rebased to retriggger tests.

mnonnenmacher commented:
> @mnonnenmacher I've rebased to retriggger tests.

That won't help as you need #4929.

sschuberth commented:
> > @mnonnenmacher I've rebased to retriggger tests.
> 
> That won't help as you need #4929.

Thanks. I've merged your and rebased mine once more.

sschuberth commented on scanner/src/main/kotlin/ScanResultsStorage.kt:
```diff
@@ -305,12 +306,10 @@ abstract class ScanResultsStorage : PackageBasedScanStorage {
-        }
-
-        if (scanResults.isEmpty()) return Success(scanResults)
+        val scanResults = readInternal(pkg.id)
```
I'll consider this for a follow-up PR.

sschuberth replied:
See https://github.com/oss-review-toolkit/ort/pull/4930/commits/0d968326e8ac5042ec336d2c457942a5bdc6102c.
